### PR TITLE
Support bloom filters in lists/nested columns

### LINKED ISSUE #98: Are bloom filters supported on LIST types?
Hi there 👋 

Firstly, thank you for this amazing library!

I'm curious to know [how to add bloom filters](https://github.com/LibertyDSNP/parquetjs#adding-bloom-filters) to `LIST` types.

For example, given this schema:

```javascript
{
  querystring: {
    type: "LIST",
    fields: {
      list: {
        repeated: true,
        fields: {
          element: {
            fields: {
              key: { type: "UTF8" },
              value: { type: "UTF8" }
            }
          }
        }
      }
    }
  }
}
```

How do you add a bloom filter for the `querystring.list.element.key` field?

```javascript
[
  {
    column: "querystring.list.element.key",
    numDistinct: 100
  }
]
```

I assume the above won't work? (Sorry in advance if that literally is how you do it!)

Thanks in advance!

### PR DESCRIPTION
**Pull request by shannonwells**
Problem
=======
Could not create and read bloom filters for a nested list.

Solution
========
For bloom filters, treat column names as the full path and not just the first part of the path in the schema.

Closes #98 

In the course of implementing this I ran across a couple of issues which are added (#104 #103 )

with @rlaferla 

Change summary:
---------------
* Modify the writer to use the full column path
* Then bloom filter options for a nested column/list can be specified by the comma-separated full path.

Steps to Verify:
----------------
1. Tests new and old should all pass.

### CODE REVIEW

File: test/bloomFilterIntegration.ts [hunk 1]
```diff
@@ -1,155 +1,257 @@
...
+        "15n is included in quantity filter"
+      );
+      assert.isFalse(
+        await splitBlockBloomFilter.check(BigInt(100)),
+        "100n is NOT included in quantity filter"
+      );
+    });
 
-    bloomFilters = await reader.getBloomFiltersFor(["name", "quantity"]);
+    it('writes bloom filters for stock,warehouse', async () => {
```
Comment (shannonwells) [inline test/bloomFilterIntegration.ts:188]: test a plain nested column bloom filter.

File: test/bloomFilterIntegration.ts [hunk 2]
```diff
@@ -1,155 +1,257 @@
...
+        await splitBlockBloomFilter.check(Buffer.from('f')),
+        "f should be in the warehouse filter"
+      );
+      assert.isFalse(
+        await splitBlockBloomFilter.check(Buffer.from('foo')),
+        "foo should not be in the warehouse filter"
+      );
+    })
   });
+  describe("a simple schema with a nested list", () => {
```
Comment (shannonwells) [inline test/bloomFilterIntegration.ts:204]: tests for a bloom filter for a nested list column.

File: lib/writer.ts [hunk 3]
```diff
@@ -259,11 +267,19 @@ export class ParquetEnvelopeWriter {
...
     this.rowGroups.forEach(group => {
       group.columns.forEach(column => {
-        const columnName = column.meta_data?.path_in_schema[0];
-        if (!columnName || columnName in this.bloomFilters === false) return;
+        if (!column.meta_data) {
+          return
+        }
+        if (!column.meta_data.path_in_schema.length) {
+          return
+        }
```
Comment (wilwade) [inline lib/writer.ts:275]: I believe this is equivalent with optional chaining
```suggestion
        if (!column.meta_data?.path_in_schema.length) {
          return
        }
```

**Review by wilwade — APPROVED** [review]
One golfing suggestion, but otherwise it is great.

- Local build
- Local pack and install in a bad setup I had

All works!

**Pull request merged**